server/src/test/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdaterTests.java

```diff
@@ -151,16 +156,68 @@ public void testDeprecationLogging() {
         assertSettingDeprecationsAndWarnings(new Setting<?>[] { deprecatedSetting });
 
         final Settings toApplyUnset = Settings.builder().putNull("logger.org.elasticsearch").build();
-        final ClusterState afterUnset = settingsUpdater.updateSettings(afterDebug, toApplyUnset, Settings.EMPTY);
+        final ClusterState afterUnset = settingsUpdater.updateSettings(afterDebug, toApplyUnset, Settings.EMPTY, logger);
         assertSettingDeprecationsAndWarnings(new Setting<?>[] { deprecatedSetting });
 
         // we also check that if no settings are changed, deprecation logging still occurs
-        settingsUpdater.updateSettings(afterUnset, toApplyUnset, Settings.EMPTY);
+        settingsUpdater.updateSettings(afterUnset, toApplyUnset, Settings.EMPTY, logger);
         assertSettingDeprecationsAndWarnings(new Setting<?>[] { deprecatedSetting });
     }
 
+    public void testUpdateWithUnknownAndSettings() {
+        runUpdateWithUnknownAndInvalidSettingTest(builder -> builder::persistentSettings, MetaData::persistentSettings);
+        runUpdateWithUnknownAndInvalidSettingTest(builder -> builder::transientSettings, MetaData::transientSettings);
+    }
+
+    private void runUpdateWithUnknownAndInvalidSettingTest(
+        final Function<MetaData.Builder, Function<Settings, MetaData.Builder>> function,
+        final Function<MetaData, Settings> settingsToTest) {
+        final Setting<String> dynamicSetting = Setting.simpleString("dynamic.setting", Property.Dynamic, Property.NodeScope);
+        final Setting<String> invalidSetting = Setting.simpleString(
+                "invalid.setting",
+                (value, settings) -> {
+                    throw new IllegalArgumentException("invalid");
+                },
+                Property.NodeScope);
+        final Settings settings = Settings.builder().put("invalid.setting", "value").put("unknown.setting", "value").build();
+
+        final Set<Setting<?>> knownSettings =
+                Stream.concat(
+                        ClusterSettings.BUILT_IN_CLUSTER_SETTINGS.stream(),
+                        Stream.of(dynamicSetting, invalidSetting))
+                        .collect(Collectors.toSet());
+        final ClusterSettings clusterSettings = new ClusterSettings(settings, knownSettings);
+        clusterSettings.addSettingsUpdateConsumer(dynamicSetting, s -> {});
+        final SettingsUpdater settingsUpdater = new SettingsUpdater(clusterSettings);
+        final ClusterState clusterState =
+                ClusterState
+                        .builder(new ClusterName("cluster"))
+                        .metaData(function.apply(MetaData.builder()).apply(settings).build())
+                        .build();
+        final Settings toApply = Settings.builder().put("dynamic.setting", "value").build();
+        final boolean applyTransient = randomBoolean();
+        final ClusterState clusterStateAfterUpdate;
+        if (applyTransient) {
```

Comment:
I took me quite a while to understand what's going on - on the one hand the persistent vs transient question is handled by parameters to this method (please rename the `function` parameter if we keep it ;)) and on the other hand we have a boolean here that deals with it directly which may be inconsistent. On top of it all, the rest API and the transport layer allows changing both at once - something we don't test. I think the functional programming got a bit out of hand here, probably due to multiple iterations. Can we have a straight forward test that works with both persistent and transient (randomly combining them)?

Reply:
I pushed d2bc7fd016f4883a0363b6e28ebcb2639711cd66.